File: problems/#96/v2.cpp
class Solution {
public:
    vector<int> dp;

    int numTrees(int n) {
        dp = vector<int>(n + 1, -1);
        dp[0] = 1;
        return solve(n);
    }

    int solve(int n) {
        if (dp[n] != -1) {
            return dp[n];
        }
        int ret = 0;
        for (int i = 1; i <= n; ++i) {
            ret += solve(i - 1) * solve(n - i);
        }
        return dp[n] = ret;
    }
};

File: problems/#234/v2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        return self.solve([head], head)

    def solve(self, curr: list[ListNode], node: Optional[ListNode]) -> bool:
        if node == None:
            return True
        ret = self.solve(curr, node.next)
        if (not ret) or curr[0].val != node.val:
            return False
        curr[0] = curr[0].next
        return True
